arch/ARC/ARCDisassembler.c

```diff
@@ -0,0 +1,473 @@
+}
+
+
+static DecodeStatus DecodeMoveHRegInstruction(MCInst *Inst, uint64_t Insn,
+					      uint64_t Address,
+					      const void *Decoder)
+{
+	;
+
+	uint64_t H = fieldFromInstruction_8(Insn, 5, 3) |
+		  (fieldFromInstruction_8(Insn, 0, 2) << 3);
+	uint64_t G = fieldFromInstruction_8(Insn, 8, 3) |
+		  (fieldFromInstruction_8(Insn, 3, 2) << 3);
+
+	if (MCDisassembler_Success != DecodeRegisterOrImm(Inst, Address, 
+									Decoder, G, 0))
+		return MCDisassembler_Fail;
+
+	return DecodeRegisterOrImm(Inst, Address, Decoder, H, Insn >> 16u);
+}
+
+static DecodeStatus DecodeCCRU6Instruction(MCInst *Inst, uint64_t Insn,
+					   uint64_t Address,
+					   const void *Decoder)
+{
+	unsigned DstB;
+	;
+	DstB = decodeBField(Insn);
+	if (DecodeGPR32RegisterClass(Inst, DstB, Address, Decoder) == MCDisassembler_Fail) {
+		return MCDisassembler_Fail;
+	}
+
+	uint64_t U6Field = fieldFromInstruction_8(Insn, 6, 6);
+	MCOperand_CreateImm0(Inst, (U6Field));
+	uint64_t CCField = fieldFromInstruction_8(Insn, 0, 4);
+	MCOperand_CreateImm0(Inst, (CCField));
+	return MCDisassembler_Success;
+}
+
+static DecodeStatus DecodeSOPwithRU6(MCInst *Inst, uint64_t Insn,
+				     uint64_t Address, const void *Decoder)
+{
+	unsigned DstB = decodeBField(Insn);
+	if (DecodeGPR32RegisterClass(Inst, DstB, Address, Decoder) == MCDisassembler_Fail) {
+		return MCDisassembler_Fail;
+	}
+
+	uint64_t U6 = fieldFromInstruction_8(Insn, 6, 6);
+	MCOperand_CreateImm0(Inst, (U6));
+	return MCDisassembler_Success;
+}
+
+static DecodeStatus DecodeSOPwithRS12(MCInst *Inst, uint64_t Insn,
+				      uint64_t Address, const void *Decoder)
+{
+	unsigned DstB = decodeBField(Insn);
+	if (DecodeGPR32RegisterClass(Inst, DstB, Address, Decoder) == MCDisassembler_Fail) {
+		return MCDisassembler_Fail;
+	}
+
+	uint64_t Lower = fieldFromInstruction_8(Insn, 6, 6);
+	uint64_t Upper = fieldFromInstruction_8(Insn, 0, 5);
+	uint64_t Sign = fieldFromInstruction_8(Insn, 5, 1) ? -1 : 1;
+	uint64_t Result = Sign * ((Upper << 6) + Lower);
+	MCOperand_CreateImm0(Inst, (Result));
+	return MCDisassembler_Success;
+}
+
+static DecodeStatus getInstruction(MCInst *Instr, uint64_t *Size, const uint8_t *Bytes,
+			    size_t BytesLen, uint64_t Address, SStream *cStream)
+{
+	DecodeStatus Result;
+	if (BytesLen < 2) {
+		*Size = 0;
+		return MCDisassembler_Fail;
+	}
+	uint8_t DecodeByte = (Bytes[1] & 0xF7) >> 3;
+	// 0x00 -> 0x07 are 32-bit instructions.
+	// 0x08 -> 0x1F are 16-bit instructions.
+	if (DecodeByte < 0x08) {
+		// 32-bit instruction.
+		if (BytesLen < 4) {
+			// Did we decode garbage?
+			*Size = 0;
+			return MCDisassembler_Fail;
+		}
+		if (BytesLen >= 8) {
+			// Attempt to decode 64-bit instruction.
+			uint64_t Insn64;
+			if (!readInstruction64(Bytes, BytesLen, Address, Size, &Insn64))
+				return MCDisassembler_Fail;
+			Result = decodeInstruction(DecoderTable64, Instr,
+						   Insn64, Address, NULL);
+			if (MCDisassembler_Success == Result) {
+				;
```

Comment:
Could you an issue please that we should also covert the `LLVM_DEBUG` stuff?
It seems to become more and more common to use.
No need to fix it here, but it nice to track.

Reply:
LLVM_DEBUG is already present in the patches